Replace multiple locks in GBPTree with single simple lock

burqen opened the pull request.

Commit:
Replace multiple locks in GBPTree

... with single simple lock.

Commit:
Nicer sleep

Commit:
Add possibility to grab cleaner and writer lock atomicly

tinwelint commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GBPTreeLock.java:
```diff
@@ -0,0 +1,119 @@
+            }
+            newState = currentState & ~targetLockBit;
+        }
+        while ( !UnsafeUtil.compareAndSwapLong( this, stateOffset, currentState, newState ) );
```
I'd guess there _could_ be an `unpark` somewhere, but since this isn't a highly contented lock at all perhaps there's no need to bother, is there?

burqen replied:
I agree that we don't need to bother with that, since the cost of making the code more complex is higher.

tinwelint commented on community/index/src/test/java/org/neo4j/index/internal/gbptree/GBPTreeLockTest.java:
```diff
@@ -0,0 +1,189 @@
+    {
+        try
+        {
+            future.get( 200, TimeUnit.MILLISECONDS );
```
I'd prefer `OtherThreadRule` instead. With it you can wait for the code to get into wait or block in specific part of code. Regardless of how long this timeout is there's a chance of flakyness and best case it waits 200ms which is unnecessary and accumulates to a non-neglectible time if there are multiple tests.

tinwelint replied:
Hmm, that rule exists in neo4j-kernel. Could be a separate PR for moving it to something like neo4j-common and making use of it.

burqen commented:
... and also need testing for two threads racing for same lock / overlapping lock.

Commit:
Move Race to common

Commit:
Fix bug in race where timeout on contestant would not throw

If last contestant exceed waiting limit, no exception would be
thrown because check was done in beginning of loop.

Commit:
Race test for GBPTreeLock

Make sure only one contestant can acquire overlapping locks
